File: ios/NativeSigner/Screens/NetworkDetails.swift
//
//  NetworkDetails.swift
//  NativeSigner
//
//  Created by Alexander Slesarev on 13.10.2021.
//

import SwiftUI

struct NetworkDetails: View {
    let content: MNetworkDetails
    let navigationRequest: NavigationRequest
    var body: some View {
        ZStack {
            VStack {
                VStack(alignment: .leading) {
                    NetworkCard(title: content.title, logo: content.logo)
                    HStack {
                        Localizable.networkName.text
                        Text(content.name)
                    }
                    HStack {
                        Localizable.base58Prefix.text
                        Text(String(content.base58prefix))
                    }
                    HStack {
                        Localizable.decimals.text
                        Text(String(content.decimals))
                    }
                    HStack {
                        Localizable.unit.text
                        Text(content.unit)
                    }
                    HStack {
                        Localizable.genesisHash.text
                        Text(content.genesisHash.formattedAsString)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                    HStack {
                        Localizable.verifierCertificateAlt.text.fixedSize(horizontal: false, vertical: true)
                        switch content.currentVerifier.ttype {
                        case "general":
                            Localizable.general.text
                        case "custom":
                            Identicon(identicon: content.currentVerifier.details.identicon)
                            VStack {
                                Localizable.custom.text
                                Text(content.currentVerifier.details.publicKey)
                                    .fixedSize(horizontal: false, vertical: true)
                                Text(Localizable.encryption(content.currentVerifier.details.encryption))
                            }
                        case "none":
                            Localizable.none.text
                        default:
                            Localizable.unknown.text
                        }
                    }
                }
                Localizable.metadataAvailable.text
                ScrollView {
                    LazyVStack {
                        ForEach(content.meta, id: \.metaHash) { metaEntry in
                            Button(
                                action: {
                                    navigationRequest(.init(action: .manageMetadata, details: metaEntry.specsVersion))
                                },
                                label: {
                                    MetadataCard(meta: metaEntry)
                                }
                            )
                        }
                    }
                }
                Spacer()
            }.padding()
        }
    }
}

// struct NetworkDetails_Previews: PreviewProvider {
// static var previews: some View {
// NetworkDetails()
// }
// }

File: android/app/src/main/java/io/parity/signer/screens/NetworkDetails.kt
package io.parity.signer.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import io.parity.signer.components.Identicon
import io.parity.signer.components.MetadataCard
import io.parity.signer.components.NetworkCard
import io.parity.signer.uniffi.Action
import io.parity.signer.uniffi.MNetworkDetails
import io.parity.signer.uniffi.MscNetworkInfo

@Composable
fun NetworkDetails(
	networkDetails: MNetworkDetails,
	button: (Action, String) -> Unit
) {
	Column {
		NetworkCard(
			network = MscNetworkInfo(
				networkTitle = networkDetails.title,
				networkLogo = networkDetails.logo
			)
		)

		Row {
			Text("Network name:")
			Text(networkDetails.name)
		}
		Row {
			Text("base58 prefix:")
			Text(networkDetails.base58prefix.toString())
		}
		Row {
			Text("decimals:")
			Text(networkDetails.decimals.toString())
		}
		Row {
			Text("unit:")
			Text(networkDetails.unit)
		}
		Row {
			Text("genesis hash:")
			Text(networkDetails.genesisHash)
		}
		Row {
			Text("Verifier certificate:")
			when (networkDetails.currentVerifier.ttype) {
				"general" -> {
					Text("general")
				}
				"custom" -> {
					Row {
						Identicon(identicon = networkDetails.currentVerifier.details.identicon)
						Column {
							Text("custom")
							Text(
								networkDetails.currentVerifier.details.publicKey
							)
							Text("encryption: " + networkDetails.currentVerifier.details.encryption)
						}
					}
				}
				"none" -> {
					Text("none")
				}
				else -> {
					Text("unknown!")
				}
			}
		}
		Text("Metadata available:")
		LazyColumn {
			items(networkDetails.meta.size) { index ->
				val metadataRecord = networkDetails.meta[index]
				Row(
					Modifier.clickable {
						button(
							Action.MANAGE_METADATA,
							metadataRecord.specsVersion
						)
					}
				) {
					MetadataCard(metadataRecord)
				}
			}
		}
	}
}
